### PR TITLE
[multistage] Return empty block if MailboxRecv nodes have no paired MailboxSender

### PR DESCRIPTION
**Pull request by xiangfu0**
Fix for https://github.com/apache/pinot/issues/11187

In the case of multi-stage query, intermediate mailboxRecv may connected to a leaf with empty table, which has no servers serving any segments, hence no mailbox sender.

In this case, the MailboxRecv can directly return an empty block.

### CODE REVIEW

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/BaseMailboxReceiveOperator.java [hunk 1]
```diff
@@ -59,13 +59,15 @@ public BaseMailboxReceiveOperator(OpChainExecutionContext context, RelDistributi
...
     MailboxMetadata senderMailBoxMetadatas =
         context.getStageMetadata().getWorkerMetadataList().get(workerId).getMailBoxInfosMap().get(senderStageId);
-    Preconditions.checkState(senderMailBoxMetadatas != null && !senderMailBoxMetadatas.getMailBoxIdList().isEmpty(),
-        "Failed to find mailbox for stage: %s",
-        senderStageId);
-    _mailboxIds = MailboxIdUtils.toMailboxIds(requestId, senderMailBoxMetadatas);
-    _mailboxes = _mailboxIds.stream()
-        .map(mailboxId -> _mailboxService.getReceivingMailbox(mailboxId))
-        .collect(Collectors.toCollection(ArrayDeque::new));
+    if (senderMailBoxMetadatas != null && !senderMailBoxMetadatas.getMailBoxIdList().isEmpty()) {
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/BaseMailboxReceiveOperator.java:62]: Preconditions.checkState((senderMailBoxMetadatas != null && !senderMailBoxMetadatas.getMailBoxIdList().isEmpty()) || expectEmptySendingMailbox) ...

and set the expectEmptySendingMailbox from MailboxAssignmentVisitor when previous stage has no server assigned. ?

Comment (walterddr) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/BaseMailboxReceiveOperator.java:62]: discussed offline. this should be allowed. instead we should make a empty server check on WorkerManager to ensure that no empty server plan for intermediate stage is allowed

Comment (xiangfu0) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/BaseMailboxReceiveOperator.java:62]: Need to modify some tests for this behavior.

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/DispatchablePlanFragment.java [hunk 2]
```diff
@@ -46,17 +47,22 @@ public class DispatchablePlanFragment {
   // used for passing custom properties to build StageMetadata on the server.
   private final Map<String, String> _customProperties;
 
+  // Used for passing unavailable segments to generate partial results.
+  private final Map<String, Set<String>> _tableToUnavailableSegments;
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/DispatchablePlanFragment.java:51]:
```suggestion
  private final Map<String, Collection<String>> _tableToUnavailableSegments;
```

File: pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java [hunk 3]
```diff
@@ -129,6 +130,11 @@ private void assignWorkersToLeafFragment(PlanFragment fragment, DispatchablePlan
         Preconditions.checkState(tableTypeToSegmentListMap.put(tableType, serverEntry.getValue()) == null,
             "Entry for server {} and table type: {} already exist!", serverEntry.getKey(), tableType);
       }
+
+      // attach unavailable segments to metadata
+      if (!routingTable.getUnavailableSegments().isEmpty()) {
+        metadata.addTableToUnavailableSegmentsMap(logicalTableName, routingTable.getUnavailableSegments());
+      }
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java:137]: cant we simply attach to `DispatchablePlanContext` this way no need to change anything in fragment

**Review by walterddr — COMMENTED** [review]
looks good to me

File: pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java (hunk 3 above)
Comment (xiangfu0) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java:137]: In the last step of planner, we need to construct DispatchableSubPlan from PlanFragment subPlanRoot and  DispatchablePlanContext dispatchablePlanContext
```
private static DispatchableSubPlan finalizeDispatchableSubPlan(PlanFragment subPlanRoot,
      DispatchablePlanContext dispatchablePlanContext) {
    return new DispatchableSubPlan(dispatchablePlanContext.getResultFields(),
        dispatchablePlanContext.constructDispatchablePlanFragmentList(subPlanRoot),
        dispatchablePlanContext.getTableNames());
  }
```

So if we don't do it here, then we need to add a new field into `DispatchableSubPlan`

Comment (walterddr) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java:137]: yeah adding a new field to dispatchable subplan is what i meant. 

e.g. 
1. `WorkerManager` here attach all unavailable segments to `dispatchablePlanContext._unavailableSegmentMap`
2. `PinotDispatchPlanner#finalizeDispatchableSubPlan` copies the map to `DispachableSubPlan._unavailableSegmentMap`

there's no need for planFragment to know this as they cannot do anything anyway. right?

Comment (xiangfu0) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java:137]: make sense.

**Review by walterddr — APPROVED** [review]

File: pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java [hunk 4]
```diff
@@ -212,6 +218,12 @@ private void assignWorkersToIntermediateFragment(PlanFragment fragment, Dispatch
     } else {
       serverInstances = fetchServersForIntermediateStage(tableNames);
     }
+    if (serverInstances.isEmpty()) {
+      LOGGER.warn("[RequestId: {}] No server instance found for intermediate stage for tables: {}",
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java:222]: logger error?

**Pull request merged**